dashboard/config/initializers/mysql_avoid_utf8_alias_in_schema_dump.rb

```diff
@@ -0,0 +1,49 @@
+# In MySQL 5.7 and earlier, `utf8` is an alias for `utf8mb3`, the three-byte
+# variant of the unicode character set. Starting with MySQL 8, `utf8` is now an
+# alias for `utf8mb4`, the four-byte variant.
+#
+# To help ease our transition from 5.7 to 8, avoid the use of the alias
+# entirely and instead use the appropriate explicit value for the current
+# verison.
+#
+# TODO infra: once we have fully upgraded to MySQL 8+, we can remove this.
+#
+# See https://dev.mysql.com/doc/refman/8.0/en/charset-unicode-utf8.html
+module MysqlAvoidUtf8AliasInSchemaDump
+  # Reimplement ActiveRecord::SchemaDumper::format_options to replace instances
+  # of the `utf8` alias with the version-specific value it is aliasing.
+  # See https://github.com/rails/rails/blob/v6.1.7.3/activerecord/lib/active_record/schema_dumper.rb#L276-L278
+  def format_options(options)
+    options.map do |key, value|
+      if key == :charset && value == 'utf8'
+        value = current_mysql_version < 8 ? 'utf8mb3' : 'utf8mb4'
+      end
+
+      if key == :collation && value == 'utf8_unicode_ci'
+        value = current_mysql_version < 8 ? 'utf8mb3_unicode_ci' : 'utf8mb4_unicode_ci'
+      end
+
+      "#{key}: #{value.inspect}"
+    end.join(', ')
+  end
```

Comment:
```
My own suggestion based on my own coding style.  


def format_options(options)
  charset, collation = charset_and_collation_values

  options.map do |key, value|
    value = case key
            when :charset then charset
            when :collation then collation
            else value
            end

    "#{key}: #{value.inspect}"
  end.join(', ')
end

def charset_and_collation_values
  if current_mysql_version < 8
    ['utf8mb3', 'utf8mb3_unicode_ci']
  else
    ['utf8mb4', 'utf8mb4_unicode_ci']
  end
end


```

Reply:
I see where you're going with this; it's certainly more extensible than my approach, and if this code was intended to live for anything longer than as short a time as possible, I think that would be valuable. But I do personally find this harder to quickly understand at a glance, and for this kind of short-term transitional code, I think that kind of simple readability wins out over strict adherence to the principles of isolation.

That being said, I do agree that there's a bit too much complexity in my `map` block. What do you think about something like,

```ruby
def format_options(options)
  options.map do |key, value|
    value = resolve_alias_for_mysql_version(key, value)
    "#{key}: #{value.inspect}"
  end.join(', ')
end

def resolve_alias_for_mysql_version(key, alias)
  if key == :charset && value == 'utf8'
    return current_mysql_version < 8 ? 'utf8mb3' : 'utf8mb4'
  end

  if key == :collation && value == 'utf8_unicode_ci'
    return current_mysql_version < 8 ? 'utf8mb3_unicode_ci' : 'utf8mb4_unicode_ci'
  end

  return value
end
```